### PR TITLE
[GEOS-11294] map ml wms vector representation

### PR DESCRIPTION
**Pull request by turingtestfail**
# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [ ] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [ ] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [ ] Bug fixes and small new features are presented as a single commit.
- [ ] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLLayerConfigurationPanel.java [hunk 1]
```diff
@@ -79,6 +81,19 @@ public MapMLLayerConfigurationPanel(final String panelId, final IModel<LayerInfo
         CheckBox useTiles = new CheckBox("useTiles", useTilesModel);
         add(useTiles);
 
+        // add the checkbox to select features or not
+        MapModel<Boolean> useFeaturesModel =
+                new MapModel<>(
+                        new PropertyModel<MetadataMap>(model, "resource.metadata"),
+                        "mapml.useFeatures");
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLLayerConfigurationPanel.java]: Fixed

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLFeatureUtil.java [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+        }
+        return mapml;
+    }
+
+    /**
+     * @param requestCRS the CRS requested by the client
+     * @param layerInfo metadata for the feature class
+     * @return
+     */
+    public static Set<Meta> deduceProjectionAndExtent(
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLFeatureUtil.java]: Fixed

File: src/extension/mapml/src/test/java/org/geoserver/mapml/MapMLWMSFeatureTest.java [hunk 3]
```diff
@@ -0,0 +1,162 @@
+package org.geoserver.mapml;
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/extension/mapml/src/test/java/org/geoserver/mapml/MapMLWMSFeatureTest.java:5]: Fixed

File: src/extension/mapml/src/test/java/org/geoserver/mapml/MapMLWMSFeatureTest.java [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+                        "EPSG:3857",
+                        null);
+
+        assertTrue(
+                "MapML response contains an exception due to multiple feature types",
+                requestResponse
+                        .response
+                        .getContentAsString()
+                        .contains(
+                                "MapML WMS Feature format does not support Multiple Feature Type output."));
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/extension/mapml/src/test/java/org/geoserver/mapml/MapMLWMSFeatureTest.java]: Fixed

File: src/extension/mapml/src/test/java/org/geoserver/mapml/MapMLWMSFeatureTest.java [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+                            + MapMLConstants.MAPML_MIME_TYPE
+                            + "&SERVICE=WMS&VERSION=1.3.0"
+                            + "&REQUEST=GetMap"
+                            + "&SRS="
+                            + srs
+                            + "&BBOX=0,0,1,1"
+                            + "&WIDTH=150"
+                            + "&HEIGHT=150"
+                            + "&format_options="
+                            + MapMLConstants.MAPML_FEATURE_FORMAT_OPTIONS
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/extension/mapml/src/test/java/org/geoserver/mapml/MapMLWMSFeatureTest.java]: Raster check implemented

File: src/extension/mapml/src/test/java/org/geoserver/mapml/MapMLWMSFeatureTest.java [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+            request.addPreferredLocale(locale);
+        }
+        request.setMethod("GET");
+        request.setContent(new byte[] {});
+        MockHttpServletResponse response = dispatch(request, "UTF-8");
+        MockRequestResponse result = new MockRequestResponse(request, response);
+        return result;
+    }
+
+    private static class MockRequestResponse {
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/extension/mapml/src/test/java/org/geoserver/mapml/MapMLWMSFeatureTest.java]: Removed

File: src/extension/mapml/src/test/java/org/geoserver/mapml/MapMLWMSTest.java [hunk 7]
```diff
@@ -226,6 +228,88 @@ public void testMapMLSingleLayer() throws Exception {
...
+                        "EPSG:3857",
+                        null);
+        StringReader readerOneNotFeatures =
+                new StringReader(requestResponseOneNotFeatures.response.getContentAsString());
+        Mapml mapmlOneNotFeatures = null;
+        try {
+            mapmlOneNotFeatures = encoder.decode(readerOneNotFeatures);
+        } catch (DataBindingException e) {
+            fail("MapML response is not valid XML");
+        }
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/extension/mapml/src/test/java/org/geoserver/mapml/MapMLWMSTest.java]: Created

**Comment by prushforth** [comment]
One quick comment (I'm still looking, will get back with more as and when appropriate).

When “Use Features” is checked/ true, the `tref` attribute of the map-link for the layer contains the “`format_options=mapmlfeatures:true`” -- this is good!

E.g. `http://localhost:8080/geoserver/tiger/wms?format_options=mapmlfeatures:true&request=GetMap&crs=MapML:OSMTILE&bbox={xmin},{ymin},{xmax},{ymax}&format=text/mapml&language=en&version=1.3.0&transparent=true&service=WMS&layers=poi&width={w}&styles=&height={h}`

The default number of decimal places for a layer in WFS settings is 0, which leads to unusable gcrs coordinates.    This is perhaps a documentation issue more than a functional problem.  I note that you can control the number of decimal places with this setting, so that’s good.

The dynamic content that is fetched by the above map-link tref contains content like that below.  **Note that there is a discrepancy between the coordinates encoded in the `<map-coordinates>` elements and the coordinate system declared via `<map-meta>` below.**  The declaration says ‘pcrs’, whereas the coordinates are encoded as gcrs.  Generally, the coordinates should be encoded in the same coordinate system that was requested via the URL, in this case that should be pcrs/ meters.  

Also, note that the response below contains several `<map-link rel=”alternate” projection=”...”>`  links.  These are strictly not necessary in the templated responses, as they are ignored and just clutter things up.  We don’t want recursively requested requests to happen, and I think the client is going to have to actively guard against that in the future, but up to now no special code is in place to prevent it IIRC.

```html
    <map-meta name="projection" content="OSMTILE"></map-meta>
    <map-meta name="extent"
        content="top-left-easting=-8238959.40,top-left-northing=4971112.17,bottom-right-easting=-8237812.69,bottom-right-northing=4969300.12"></map-meta>
    <map-meta name="cs" content="pcrs"></map-meta>    <!-- this is the correct declaration -->

<map-link href="http://localhost:8080/geoserver/tiger/wms?LANGUAGE=en&amp;CRS=MapML:OSMTILE&amp;FORMAT=text/mapml&amp;STYLES=&amp;WIDTH=588&amp;HEIGHT=911&amp;LAYERS=poi&amp;REQUEST=GetMap&amp;FORMAT_OPTIONS=mapmlfeatures:true&amp;BBOX=-8239245.213801639,4968578.821850099,-8237840.683406899,4970754.888499092&amp;SRSNAME=MapML:OSMTILE&amp;VERSION=1.3.0&amp;SERVICE=WMS&amp;TRANSPARENT=true&amp;Language=en" rel="alternate" projection="OSMTILE"></map-link>


    <map-feature id="poi.1" class="poi">
        <map-geometry>
            <map-point>
                <map-coordinates>-74.010461 40.707588</map-coordinates>  <!-- this encoding is gcrs, which doesn't follow the declaration above. Should be in projected / meters. -->
            </map-point>
        </map-geometry>
        <map-properties>
            <table xmlns="http://www.w3.org/1999/xhtml">
...
            </table>
        </map-properties>
    </map-feature>
```

**Comment by turingtestfail** [comment]
> One quick comment (I'm still looking, will get back with more as and when appropriate).
> 
> When “Use Features” is checked/ true, the `tref` attribute of the map-link for the layer contains the “`format_options=mapmlfeatures:true`” -- this is good!
> 
> E.g. `http://localhost:8080/geoserver/tiger/wms?format_options=mapmlfeatures:true&request=GetMap&crs=MapML:OSMTILE&bbox={xmin},{ymin},{xmax},{ymax}&format=text/mapml&language=en&version=1.3.0&transparent=true&service=WMS&layers=poi&width={w}&styles=&height={h}`
> 
> The default number of decimal places for a layer in WFS settings is 0, which leads to unusable gcrs coordinates. This is perhaps a documentation issue more than a functional problem. I note that you can control the number of decimal places with this setting, so that’s good.
> 
> The dynamic content that is fetched by the above map-link tref contains content like that below. **Note that there is a discrepancy between the coordinates encoded in the `<map-coordinates>` elements and the coordinate system declared via `<map-meta>` below.** The declaration says ‘pcrs’, whereas the coordinates are encoded as gcrs. Generally, the coordinates should be encoded in the same coordinate system that was requested via the URL, in this case that should be pcrs/ meters.
> 
> Also, note that the response below contains several `<map-link rel=”alternate” projection=”...”>` links. These are strictly not necessary in the templated responses, as they are ignored and just clutter things up. We don’t want recursively requested requests to happen, and I think the client is going to have to actively guard against that in the future, but up to now no special code is in place to prevent it IIRC.
> 
> ```
>     <map-meta name="projection" content="OSMTILE"></map-meta>
>     <map-meta name="extent"
>         content="top-left-easting=-8238959.40,top-left-northing=4971112.17,bottom-right-easting=-8237812.69,bottom-right-northing=4969300.12"></map-meta>
>     <map-meta name="cs" content="pcrs"></map-meta>    
> 
> <map-link href="http://localhost:8080/geoserver/tiger/wms?LANGUAGE=en&amp;CRS=MapML:OSMTILE&amp;FORMAT=text/mapml&amp;STYLES=&amp;WIDTH=588&amp;HEIGHT=911&amp;LAYERS=poi&amp;REQUEST=GetMap&amp;FORMAT_OPTIONS=mapmlfeatures:true&amp;BBOX=-8239245.213801639,4968578.821850099,-8237840.683406899,4970754.888499092&amp;SRSNAME=MapML:OSMTILE&amp;VERSION=1.3.0&amp;SERVICE=WMS&amp;TRANSPARENT=true&amp;Language=en" rel="alternate" projection="OSMTILE"></map-link>
> 
> 
>     <map-feature id="poi.1" class="poi">
>         <map-geometry>
>             <map-point>
>                 <map-coordinates>-74.010461 40.707588</map-coordinates>  
>             </map-point>
>         </map-geometry>
>         <map-properties>
>             <table xmlns="http://www.w3.org/1999/xhtml">
> ...
>             </table>
>         </map-properties>
>     </map-feature>
> ```

Heads up that I just pushed a fix to the number of decimals issue.  It was defaulting to zero but now it should default to the global setting (which in turn defaults to 8).  This matches how GML2 handles the same issue in WFS.

**Comment by prushforth** [comment]
I'm unable to get (for example) the states layer to serve up vectors in pcrs coordinates for MapML:OSMTILE.  Works well, including the bugfix for default decimals, for MapML:WGS84.  Getting non-WGS84 projections working is essential, but am I missing something?  Thanks

**Comment by turingtestfail** [comment]
Peter,
It looks like in some cases the ReprojectingFeatureCollection I introduced
for this conversion is not able to detect the CRS from the geometry,
leading it to return the coordinates in the stored CRS.  I just pushed a
fix that establishes a fallback CRS based on the stored layer configuration
that should address the issue.  Can you do a pull and test again?
Joe

**Comment by prushforth** [comment]
> Can you do a pull and test again?

ok that seems to make a big difference for me. I will proceed to look more!

Thanks!

**Comment by prushforth** [comment]
I think that when the "Use Features" MapML option is selected, that we should leave out the <map-link rel="query" ...> link.  It seems redundant and may mess somebody up.  I can take care of that, once I take a look at and understand your code.

Looking pretty good!  I note that no clipping to the bbox is happening yet, correct?  i.e. the whole layer is requested at each map movement.  

I also note that the popup for large (many attributes) attribute tables is really not great.  We'll raise an issue on the client for that.

**Comment by aaime** [comment]
@prushforth yes, handling clipping is going to be my first task once this PR is merged, while @turingtestfail dives into style translation.

**Comment by turingtestfail** [comment]
Peter,
Once the   <map-link rel="query" ...> link is removed are you OK with us
moving forward with a PR against the main repository?
Thanks,
Joe

**Comment by prushforth** [comment]
> Once the   <map-link rel="query" ...> link is removed are you OK with us
moving forward with a PR against the main repository?

Yes!

**Pull request closed without merging**